sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -44,82 +44,116 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + azimuth_offset]);
+      float azimuth_diff = next_azimuth - current_azimuth;
       azimuth_diff = azimuth_diff < 0.f ? azimuth_diff + 36000.f : azimuth_diff;
 
-      const auto current_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[current_idx + distance_offset]);
-      const auto next_pt_distance =
-        *reinterpret_cast<float *>(&input_ptr->data[next_idx + distance_offset]);
+      const float & current_distance =
+        *reinterpret_cast<const float *>(&input->data[current_data_idx + distance_offset]);
+      const float & next_distance =
+        *reinterpret_cast<const float *>(&input->data[next_data_idx + distance_offset]);
 
       if (
-        std::max(current_pt_distance, next_pt_distance) <
-          std::min(current_pt_distance, next_pt_distance) * distance_ratio_ &&
+        std::max(current_distance, next_distance) <
+          std::min(current_distance, next_distance) * distance_ratio_ &&
         azimuth_diff < 100.f) {
-        continue;
+        continue;  // Determined to be included in the same walk
       }
-      if (isCluster(input_ptr, tmp_indices)) {
-        for (const auto & tmp_idx : tmp_indices) {
-          output_modifier.push_back(
-            std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+      if (isCluster(
+            input, indices[walk_first_idx], indices[walk_last_idx],
+            walk_last_idx - walk_first_idx + 1)) {
+        for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+          auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+          *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+          output_size += output.point_step;
         }
       }
-      tmp_indices.clear();
-    }
-    if (tmp_indices.empty()) {
-      continue;
+
+      walk_first_idx = idx + 1;
     }
-    if (isCluster(input_ptr, tmp_indices)) {
-      for (const auto & tmp_idx : tmp_indices) {
-        output_modifier.push_back(
-          std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+    if (walk_first_idx > walk_last_idx) continue;
+
+    if (isCluster(
+          input, indices[walk_first_idx], indices[walk_last_idx],
+          walk_last_idx - walk_first_idx + 1)) {
+      for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+        auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+        *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+        output_size += output.point_step;
       }
     }
-    tmp_indices.clear();
   }
+
+  output.data.resize(output_size);
+
+  // Note that `input->header.frame_id` is data before converted when `transform_info.need_transform
+  // == true`
+  output.header.frame_id = !tf_input_frame_.empty() ? tf_input_frame_ : tf_input_orig_frame_;
+
+  output.height = 1;
+  output.fields = input->fields;
```

Comment:
Comming back to my previous comment about the `PointXYZI`data copy. Here we say input and output has the same fields, yet we don't copy "azimut" and "ring" fields from input.